### PR TITLE
[enh] Add fail2ban helpers

### PR DESCRIPTION
**Pull request by JimboJoe**
## Problems
To enhance applications protection against hackers/spammers, etc., we can propose helpers to ease the creating of fail2ban jails.

## Solution
Add `ynh_add_fail2ban_config` and `ynh_remove_fail2ban_config` helpers.
A successful implementation example is the [piwigo app](https://github.com/YunoHost-Apps/piwigo_ynh).

## PR Status
Working, but opinion welcome!
And should be implemented in other applications to validate its principle.

## Validation

- [ ] Principle agreement 0/2 : 
- [ ] Quick review 0/1 : 
- [ ] Simple test 0/1 : 
- [ ] Deep review 0/1 :

### CODE REVIEW

**Comment by Psycojoker** [comment]
As anyone tested those? I have no knowledge of fail2ban.

**Comment by JimboJoe** [comment]
At the moment, only the (nearly official) Piwigo application implemented it. So maybe should we wait for a couple of other successful implementations to merge it...?

**Comment by alexAubin** [comment]
As far as I know, fail2ban's conf format changed in Stretch so that might affect these helpers :s

**Comment by Psycojoker** [comment]
> At the moment, only the (nearly official) Piwigo application implemented it. So maybe should we wait for a couple of other successful implementations to merge it...?

Any update on that?

**Comment by JimboJoe** [comment]
It's currently being implemented in Wordpress, and has evolved [here](https://github.com/YunoHost-Apps/Experimental_helpers/pull/2).

**Comment by alexAubin** [comment]
About this : 

we should be careful that the fail2ban conf format changed a bit in stretch and is backward-incompatible. I don't know if these helpers are affected... but right now, the step in the stretch migration is to remove all existing fail2ban conf and regenerate it from scratch :s

**Comment by JimboJoe** [comment]
I've updated the helpers to be on-par with the ones from Experimental_helpers.
I've tested again piwigo on Stretch and they work as advertised.
I understand you had to delete all existing fail2ban jails when migrating from Jessie to Stretch, so that means existing apps won't be protected by fail2ban until their next upgrade (which I find perfectly acceptable)?

**Comment by alexAubin** [comment]
>I understand you had to delete all existing fail2ban jails when migrating from Jessie to Stretch, so that means existing apps won't be protected by fail2ban until their next upgrade

Yes indeed :s

**Comment by JimboJoe** [comment]
This helper currently [only works on Stretch](https://github.com/YunoHost-Apps/Experimental_helpers/pull/11#issuecomment-390648486)...

**Comment by zamentur** [comment]
Not a problem if it only works on Stretch, we could merged it only in stretch...

**Comment by Psycojoker** [comment]
We are now on stretch, shouldn't we merge this?

File: data/helpers.d/backend [hunk 1]
```diff
@@ -183,3 +183,76 @@ ynh_remove_fpm_config () {
...
+
+  test -n "$logpath" || ynh_die "ynh_add_fail2ban_config expects a logfile path as first argument and received nothing."
+  test -n "$failregex" || ynh_die "ynh_add_fail2ban_config expects a failure regex as second argument and received nothing."
+
+  finalfail2banjailconf="/etc/fail2ban/jail.d/$app.conf"
+  finalfail2banfilterconf="/etc/fail2ban/filter.d/$app.conf"
+  ynh_backup_if_checksum_is_different "$finalfail2banjailconf" 1
+  ynh_backup_if_checksum_is_different "$finalfail2banfilterconf" 1
+
+  sudo tee $finalfail2banjailconf <<EOF
```
Comment (Josue-T) [inline data/helpers.d/backend]: As all script all called as `root` why does we continue to use sudo ?

**Comment by maniackcrudelis** [comment]
I've updated the helpers to include the fix for jessie and the last modifications from the experimental repository.
This helper will now need #520

File: data/helpers.d/backend (hunk 1 above)
Comment (maniackcrudelis) [reply data/helpers.d/backend]: Probably because someone have forgotten to remove it.

Comment (Josue-T) [reply data/helpers.d/backend]: It could be good to clean this everywhere one day.

**Comment by JimboJoe** [comment]
Yes, but not before #520 gets merged :wink:

**Comment by alexAubin** [comment]
#520 got merged. Any update about this ? :D

**Comment by alexAubin** [comment]
Bumpz D:

**Comment by Josue-T** [comment]
I think it could be useful to have more flexibility about the config file. By example if we need to specify multiple regex or add some ingoreregex or add the "maxlines" option.

**Comment by alexAubin** [comment]
Okay guys, I'm flagging this as inactive :/ ... It'd be really nice if we could have some progress on this one way or another. I don't think we need a 100% perfect solution to deploy this, considering that in the meantime some web apps may remain vulnerable to brute force.

**Comment by maniackcrudelis** [comment]
Let's merge in 3 days if no one disagree until then.

**Comment by alexAubin** [comment]
Following today's meeting, this PR is postponed. Josue will have a look at integrating fail2ban in his apps, which requires to have several regex instead of just one. He'll iterate on the design.

File: data/helpers.d/backend [hunk 2]
```diff
@@ -183,3 +183,76 @@ ynh_remove_fpm_config () {
...
+  max_retry=${max_retry:-3}
+  ports=${ports:-http,https}
+
+  test -n "$logpath" || ynh_die "ynh_add_fail2ban_config expects a logfile path as first argument and received nothing."
+  test -n "$failregex" || ynh_die "ynh_add_fail2ban_config expects a failure regex as second argument and received nothing."
+
+  finalfail2banjailconf="/etc/fail2ban/jail.d/$app.conf"
+  finalfail2banfilterconf="/etc/fail2ban/filter.d/$app.conf"
+  ynh_backup_if_checksum_is_different "$finalfail2banjailconf" 1
+  ynh_backup_if_checksum_is_different "$finalfail2banfilterconf" 1
```
Comment (maniackcrudelis) [inline data/helpers.d/backend]: Is anyone know why is there a '1' as a second argument for this helper ?
I can't find any version of `ynh_backup_if_checksum_is_different` with a second argument.

This is crashing getopts, which is another problem to fix

**Comment by maniackcrudelis** [comment]
@YunoHost/apps

**Comment by alexAubin** [comment]
Bump @YunoHost/apps , it'd be nice to have this for 3.5 :s (in particular, @Josue-T , it was noted "Josue will have a look at integrating fail2ban in his apps, which requires to have several regex instead of just one. He'll iterate on the design." :sweat_smile: so if you happen to have time to work on this that'd be great !)

**Comment by Josue-T** [comment]
Yes, I might find some time for that in theses next 2 weeks.
If not I think we should merge this as it is because I might not find the time for the theses next month

**Comment by Josue-T** [comment]
Now I updated the helper.

Tested on [synapse](https://github.com/YunoHost-Apps/synapse_ynh/pull/106)

**Comment by maniackcrudelis** [comment]
@Josue-T that's a completely different helper !
You break getopts usage as well as usage of this helper in one line without template.

Maybe for synapse you need to use a template, but most of the apps don't need that and works perfectly with the previous version of this helper.
So, if you want to ADD a template usage for this helper, why not, but I don't think that to force to use a template is a good idea.

**Comment by Josue-T** [comment]
It's just more like the helper nginx.

I think it's the last time change the way how this helper work. And I think that probably many apps need more flexibility than what this helper provided.

**Comment by Josue-T** [comment]
If you really want to keep first idea. We can make a helper `ynh_add_fail2ban_config` with the old idea. And a other helper `ynh_add_fail2ban_template` with my idea.

**Comment by maniackcrudelis** [comment]
nginx doesn't have a standard config, each app has a different config.
You could have said that it's like php, which is true. And actually, this helper could use a template less architecture. But the helper was built over the way we were handling php by that time.

Anyway, except synapse, no app is using a template for fail2ban. And so far we don't need to add that in apps. So we should keep that helper as simple to use as we can. So, like it was.

Again, if you need a template feature for synapse with fail2ban, I agree to ADD this to the helper. But, this shouldn't break the way the helper was working so far.
Nor breaking getopts...

So, add 2 arguments to the helper, one for using a template, another for adding other variables. That's a good thing for those who could need a specific fail2ban config. But do not break the helper because you need a specific way to handle the config.

**Comment by Josue-T** [comment]
Now there are the 2 possibility

**Comment by maniackcrudelis** [comment]
LGTM, but not tested.

**Comment by maniackcrudelis** [comment]
Bump

@YunoHost/apps

**Comment by maniackcrudelis** [comment]
Let's merge in 3 days

**Comment by alexAubin** [comment]
Thanks for the review folks, finally sounds like we gonna merge this \o/ 

serfefefeveimage

**Pull request merged**